Mount: Switch to RC_TARGETING on RC input

Davidsastresas opened the pull request:
This PR addresses the issue with mixed mavlink and manual control for gimbals. Right now after each mavlink command we do, it would be in mavlink targeting mode and in order to move manually we would need to manually set it to rc_targetting. This is very confusing and counter intuitive for the user, because during regular operation it is very common to alternate mavlink commands with manual, mixing ROI with manual control, Neutral every now and then to make sense of where the vehicle is pointing to, etc.

This way the operation feels much more natural.

It also sends a MAV_CMD_DO_SET_ROI_NONE command if coming from ROI targeting. This way the GCS can understand the vehicle is no longer doing ROI, and can update the vehicle status representation in the UI.

There is a particular case that might be worth discussing, retracted mode. Maybe we could leave this mode out of this automatic behavior? I am thinking it could be easy to set the mount to retracted, and unconsciously move the manual sticks, it would not be retracted anymore. Maybe in this particular scenario we could require manual "un-retracting"? Not saying it is wrong as it is now, just wanted to open the discussion.

Thank you very much @rmackay9 !

rmackay9 committed:
AP_Mount: switch to RC_TARGETING on RC input

Davidsastresas committed:
AP_Mount_Backend.cpp: send ROI_NONE after automatic RC_TARGETING

Davidsastresas commented:
Sorry, I didn't mean to make the PR here yet, I thought I was in randy's repo:

rmackay9 commented:
Hi @Davidsastresas,

I agree that we should disable the automatic switch to RC if the gimbal is in Retracted mode.

As discussed separately I'm less sure about the proactive sending of the COMMAND_ACK message when no GCS has actually sent a command.  I think the GCSs need to be informed of the change in the mount's mode but sadly we don't have a way to do that at the moment.

Davidsastresas committed:
Revert "AP_Mount_Backend.cpp: send ROI_NONE after automatic RC_TARGETING"

This reverts commit d6f710d7b5c61a2f2e91d6ba54532decbf942dde.

Davidsastresas committed:
AP_Mount_Backend.cpp: send one gimbal_manager_status msg when control changes:

By default we are sending this message at 0.2 Hz. This is totally
fine as no more rate is needed, but whenever control changes it
is interesting to notify as soon as possible, so the rest of the
mavlink network understands the change in control as soon as
possible

Davidsastresas commented:
As per Randy's recomendation I reverted the commit about sending the ACK for ROI_NONE, we can deal with that in the future.

I added a new commit to send immediately a gimbal_manager_status message whenever the sysid/compid in control of a gimbal changes. Here is a brief demonstation of it in the ongoing QGC gimbal development work https://github.com/mavlink/qgroundcontrol/pull/10667:

https://github.com/ArduPilot/ardupilot/assets/18221398/bb5decf5-c20e-4aae-8ff5-c6cf43c4a7d4

As it is right now, the code is kind of redundant, very similar code is in send_gimbal_manager_status function. I didn't reuse that because I don't know if there is a possibility of getting the "chan" argument from this one to reuse from within the same class. If there is a different way of doing that and keeping the code more compact let me know and I will adapt accordingly.

Thanks!

rmackay9 commented on libraries/AP_Mount/AP_Mount_Backend.cpp:
```diff
@@ -382,6 +385,23 @@ MAV_RESULT AP_Mount_Backend::handle_command_do_gimbal_manager_configure(const ma
+        if (_yaw_lock) {
+            flags |= GIMBAL_MANAGER_FLAGS_YAW_LOCK;
+        }
+        const mavlink_gimbal_manager_status_t manager_status_packet{
```
It would be nice if we could instead re-use the code that we have in AP_Mount_Backend::send_gimbal_manager_status().

If we're happy with the gimbal-manager-status going out to only the channel that we received the do-gimbal-manager-configure command on then we could pass the channel into handle_command_do_gimbal_manager_configure() and reuse it.  If instead we'd like all mavlink channels to get the message then we could call gcs().send_message(MSG_GIMBAL_MANAGER_STATUS);

Davidsastresas replied:
Thank you! I just saw the other PR, sorry I just saw this now.

rmackay9 reviewed (changes requested):
Thanks again for this.

Could you remove the "revert" commit (and the original).  I'm sure you were planning to anyway but just in case.

peterbarker commented:
@Davidsastresas we merged @rmackay9 's adjustments of this one.

Closing now - thanks!
